File: templates/templates/ruby/tests/primary/flattening_test.rb
# frozen_string_literal: true
# typed: true

require_relative "../lib/openapi"
require_relative "common_helper_test"
require_relative "helper_test"

require "minitest/autorun"
require "minitest/focus"
require "rack"

module OpenApiSDK
  class TestFlattening < Minitest::Test
    def test_component_body_and_param_no_conflict
      record_test("flattening-component-body-and-param-no-conflict")

      @sdk = OpenApiSDK::SDK.new(server_url: CommonHelpers.httpbin_url)
      refute_nil(@sdk)

      obj = create_simple_object

      res = @sdk.flattening.component_body_and_param_no_conflict(simple_object: obj, param_str: "param test")
      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.res)
      assert_equal(res.res.args["paramStr"], "param test")
      assert_equal(res.res.json, obj)
    end

    def test_component_body_and_param_conflict
      record_test("flattening-component-body-and-param-conflict")

      @sdk = OpenApiSDK::SDK.new(server_url: CommonHelpers.httpbin_url)
      refute_nil(@sdk)

      obj = create_simple_object

      res = @sdk.flattening.component_body_and_param_conflict(simple_object: obj, str_: "param test")
      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.res)
      assert_equal(res.res.args["str"], "param test")
      assert_equal(res.res.json, obj)
    end

    def test_inline_body_and_param_conflict
      record_test("flattening-inline-body-and-param-conflict")

      @sdk = OpenApiSDK::SDK.new(server_url: CommonHelpers.httpbin_url)
      refute_nil(@sdk)

      res = @sdk.flattening.inline_body_and_param_conflict(
        request_body: Models::Operations::InlineBodyAndParamConflictRequestBody.new(
          str_: "body test"
        ),
        str_: "param test"
      )
      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.res)
      assert_equal(res.res.args["str"], "param test")
      assert_equal(res.res.json.str_, "body test")
    end

    def test_inline_body_and_param_no_conflict
      record_test("flattening-inline-body-and-param-no-conflict")

      @sdk = OpenApiSDK::SDK.new(server_url: CommonHelpers.httpbin_url)
      refute_nil(@sdk)

      res = @sdk.flattening.inline_body_and_param_no_conflict(
        request_body: Models::Operations::InlineBodyAndParamNoConflictRequestBody.new(
          body_str: "body test"
        ),
        param_str: "param test"
      )
      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.res)
      assert_equal(res.res.args["paramStr"], "param test")
      assert_equal(res.res.json.body_str, "body test")
    end

    def test_conflicting_params
      record_test("flattening-conflicting-params")

      @sdk = OpenApiSDK::SDK.new(server_url: CommonHelpers.httpbin_url)
      refute_nil(@sdk)

      res = @sdk.flattening.conflicting_params(str_path_parameter: "pathParam", str_query_parameter: "queryParam")
      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.res)
      assert_match(%r{/pathParam?}, res.res.url)
      assert_equal(res.res.args["str"], "queryParam")
    end

    def test_required_body_all_optional
      record_test("flattening-required-body-all-optional")

      @sdk = OpenApiSDK::SDK.new(server_url: CommonHelpers.httpbin_url)
      refute_nil(@sdk)

      res = @sdk.flattening.required_body_all_optional(
        request: Models::Shared::ObjWithOptionalProperties.new(
          opt_str: "body test",
          opt_int: 1
        )
      )
      refute_nil(res)
      assert_equal(Rack::Utils.status_code(:ok), res.http_meta.response.status)
      refute_nil(res.res)
      assert_equal(res.res.json.opt_str, "body test")
      assert_equal(res.res.json.opt_int, 1)

      assert_raises(ArgumentError, "wrong number of arguments (given 0, expected 1)") do
        @sdk.flattening.required_body_all_optional
      end
    end

  end
end

File: templates/templates/typescriptv2/tests/no-zod/flattening.test.ts
import { expect, test } from "vitest";

import { SDK } from "../index.js";

import { HTTPBIN_URL, recordTest } from "./common_helpers.js";
import { createSimpleObject } from "./primary_helpers.js";

test("Test Component Body and Param No Conflict", async () => {
  recordTest("flattening-component-body-and-param-no-conflict");

  const s = new SDK({ serverURL: HTTPBIN_URL });
  const obj = createSimpleObject();
  const res = await s.flattening.componentBodyAndParamNoConflict(
    "param test",
    obj,
  );
  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();
  expect(res.res!.args["paramStr"]).toEqual("param test");
  expect(res.res!.json).toEqual(obj);
});

test("Test Component Body and Param Conflict", async () => {
  recordTest("flattening-component-body-and-param-conflict");

  const s = new SDK({ serverURL: HTTPBIN_URL });
  const obj = createSimpleObject();
  const res = await s.flattening.componentBodyAndParamConflict(
    "param test",
    obj,
  );
  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();
  expect(res.res!.args["str"]).toEqual("param test");
  expect(res.res!.json).toEqual(obj);
});

test("Test Inline Body and Param Conflict", async () => {
  recordTest("flattening-inline-body-and-param-conflict");

  const s = new SDK({ serverURL: HTTPBIN_URL });
  const res = await s.flattening.inlineBodyAndParamConflict("param test", {
    str: "body test",
  });
  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();
  expect(res.res!.args["str"]).toEqual("param test");
  expect(res.res!.json.str).toEqual("body test");
});

test("Test Inline Body and Param No Conflict", async () => {
  recordTest("flattening-inline-body-and-param-no-conflict");

  const s = new SDK({ serverURL: HTTPBIN_URL });
  const res = await s.flattening.inlineBodyAndParamNoConflict("param test", {
    bodyStr: "body test",
  });
  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();
  expect(res.res!.args["paramStr"]).toEqual("param test");
  expect(res.res!.json.bodyStr).toEqual("body test");
});

test("Test Conflicting Params", async () => {
  recordTest("flattening-conflicting-params");

  const s = new SDK({ serverURL: HTTPBIN_URL });
  const res = await s.flattening.conflictingParams("pathParam", "queryParam");
  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();
  expect(res.res!.url).toContain("/pathParam?");
  expect(res.res!.args["str"]).toEqual("queryParam");
});

test("Test Required Body All Optional", async () => {
  recordTest("flattening-required-body-all-optional");

  const s = new SDK({ serverURL: HTTPBIN_URL });
  const res = await s.flattening.requiredBodyAllOptional({
    optStr: "body test",
    optInt: 1,
  });
  expect(res.httpMeta.response.status).toBe(200);
  expect(res.res).toBeDefined();
  expect(res.res!.json.optStr).toEqual("body test");
  expect(res.res!.json.optInt).toEqual(1);
});
